{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# 1 Quick Start\n",
    "This notebook provides instructions on how to get your vrs-python environment up and running with as few\n",
    "steps as possible, and to provide some rudimentary examples to prove it is working properly.\n",
    "\n",
    "This assumes that you have successfully performed the pre-requisite installations and setup steps detailed in [README.md](README.md).\n",
    "\n",
    "The vrs-python package has a dependency on the [biocommons seqrepo package](https://github.com/biocommons/biocommons.seqrepo). SeqRepo is used for referencing biological sequences. In this series of notebooks we will be using the publicly available SeqRepo when initializing *DataProxy* as seen below in Step 1:\n",
    "\n",
    "    seqrepo_rest_service_url = \"seqrepo+https://services.genomicmedlab.org/seqrepo\"\n",
    "\n",
    "Another dependency of vrs-python is the [biocommons hgvs package](https://github.com/biocommons/hgvs) for parsing HGVS nomenclature. The hgvs package further relies on the [biocommons uta package](https://github.com/biocommons/uta). The Universal Transcript Archive (UTA) is a Postgres database that stores transcripts aligned to sequence references. When necessary, we will define `UTA_DB_URL` in the environment using the public access version as follows:\n",
    "\n",
    "    UTA_DB_URL=\"postgresql://anonymous:anonymous@uta.biocommons.org:5432/uta/uta_20210129b\"\n",
    "\n",
    "**NOTE** The external sources for the SeqRepo and UTA repositories are **ONLY** to be used as part of this notebook series and are not meant for use in production code. Please refer to the links above and follow the directions provided on how to setup local instances."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "7f11c55b5ce145ee"
  },
  {
   "cell_type": "markdown",
   "id": "91f92a3e35bd48a1",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Step 1 - Setup Data Proxy Access\n",
    "The *DataProxy* provides access to sequence references."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "37130d69b9dbd9d1",
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:17.340604Z",
     "start_time": "2024-04-18T20:21:17.207979Z"
    }
   },
   "outputs": [],
   "source": [
    "from ga4gh.vrs.dataproxy import create_dataproxy\n",
    "seqrepo_rest_service_url = \"seqrepo+https://services.genomicmedlab.org/seqrepo\"\n",
    "seqrepo_dataproxy = create_dataproxy(uri=seqrepo_rest_service_url)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a1d5bbc3a77ff03f",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "Assert that the UTA URL is defined in the environment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "b11653c9aae4ecba",
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:17.343195Z",
     "start_time": "2024-04-18T20:21:17.341690Z"
    }
   },
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"UTA_DB_URL\"] = \"postgresql://anonymous:anonymous@uta.biocommons.org:5432/uta/uta_20210129b\""
   ]
  },
  {
   "cell_type": "markdown",
   "id": "98ab29e5ac01cc3a",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Step 2 - Setup an Allele Translator\n",
    "Now we will create and utilize an *AlleleTranslator* to test that our setup is good. The *AlleleTranslator* has the ability to translate single nucleotide variants and simple insertions/deletions forms to VRS. Notice the *AlleleTranslator* dependency on the *DataProxy*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "42bd6d6f09916724",
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:17.393606Z",
     "start_time": "2024-04-18T20:21:17.343980Z"
    }
   },
   "outputs": [],
   "source": [
    "from ga4gh.vrs.extras.translator import AlleleTranslator\n",
    "translator = AlleleTranslator(data_proxy=seqrepo_dataproxy)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88d9aba51e44ae0",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "#### Step 3 - Translate variation representations to VRS\n",
    "Now we are ready to have the *AlleleTranslator* transform our first SPDI variant representation to VRS.\n",
    "This variant can be viewed in [Clinvar](https://www.ncbi.nlm.nih.gov/clinvar/variation/2673535/?oq=2673535)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "af7a8f1509acf4ed",
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:17.525029Z",
     "start_time": "2024-04-18T20:21:17.394248Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.LK_4rOVxyEwrEpaOVd-BDFV0ocbO5vgV',\n 'type': 'Allele',\n 'digest': 'LK_4rOVxyEwrEpaOVd-BDFV0ocbO5vgV',\n 'location': {'id': 'ga4gh:SL.nA5-KovovkH-5p3LF1657nkkeWFwrInI',\n  'type': 'SequenceLocation',\n  'digest': 'nA5-KovovkH-5p3LF1657nkkeWFwrInI',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.aUiQCzCPZ2d0csHbMSbh2NzInhonSXwI'},\n  'start': 80656509,\n  'end': 80656510},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'TT'}}"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vrs_from_spdi = translator.translate_from(\"NC_000005.10:80656509:C:TT\", \"spdi\")\n",
    "vrs_from_spdi.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6fa7cb2ac050547f",
   "metadata": {
    "collapsed": false
   },
   "source": [
    "The output above is the JSON structure of an *Allele* in VRS form. You should be able to recognize the *Allele*, *SequenceLocation*, *SequenceReference* and *LiteralSequenceLocation* classes. \n",
    " \n",
    "Now we will pass the HGVS variant representation of the same variant to the *AlleleTranslator*."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "42430c680fe262c1",
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:18.708118Z",
     "start_time": "2024-04-18T20:21:17.526585Z"
    }
   },
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.LK_4rOVxyEwrEpaOVd-BDFV0ocbO5vgV',\n 'type': 'Allele',\n 'digest': 'LK_4rOVxyEwrEpaOVd-BDFV0ocbO5vgV',\n 'location': {'id': 'ga4gh:SL.nA5-KovovkH-5p3LF1657nkkeWFwrInI',\n  'type': 'SequenceLocation',\n  'digest': 'nA5-KovovkH-5p3LF1657nkkeWFwrInI',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.aUiQCzCPZ2d0csHbMSbh2NzInhonSXwI'},\n  'start': 80656509,\n  'end': 80656510},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'TT'}}"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "vrs_from_hgvs = translator.translate_from(\"NC_000005.10:g.80656510delinsTT\", \"hgvs\")\n",
    "vrs_from_hgvs.model_dump(exclude_none=True)"
   ]
  },
  {
   "cell_type": "markdown",
   "source": [
    "The VRS variant representations should be the same."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "2c6d2303b4bda87c"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "assert(vrs_from_hgvs == vrs_from_spdi)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:21:18.713821Z",
     "start_time": "2024-04-18T20:21:18.710176Z"
    }
   },
   "id": "3c91be2bdec6b4be",
   "execution_count": 6
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.13"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
