{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# 4 Exploring the AlleleTranslator\n",
    "There are four variant nomenclatures available in the vrs-python *AlleleTranslator*: SPDI, gnomad/VCF, Beacon and HGVS. In this notebook we will perform a simple Allele translation for each. We will use each of the four nomenclatures translating variants to VRS."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f3a35f19da823af8"
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Step 1 - Setup Data Proxy Access\n",
    "The *DataProxy* provides access to sequence references."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5b3ec967f1e08834"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "from ga4gh.vrs.dataproxy import create_dataproxy\n",
    "seqrepo_rest_service_url = \"seqrepo+https://services.genomicmedlab.org/seqrepo\"\n",
    "seqrepo_dataproxy = create_dataproxy(uri=seqrepo_rest_service_url)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:39.060697Z",
     "start_time": "2024-04-19T15:29:38.919390Z"
    }
   },
   "id": "4dd605526ab7227e",
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "source": [
    "Import the *AlleleTranslator* class."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "5f7ac7c602d40af7"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "from ga4gh.vrs.extras.translator import AlleleTranslator"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:39.129350Z",
     "start_time": "2024-04-19T15:29:39.061994Z"
    }
   },
   "id": "24b10ea2d6ae9b0b",
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "source": [
    "The UTA server is required in the environment since we are translating from/to HGVS."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6381d3a17866d76d"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"UTA_DB_URL\"] = \"postgresql://anonymous:anonymous@uta.biocommons.org:5432/uta/uta_20210129b\""
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:39.131648Z",
     "start_time": "2024-04-19T15:29:39.130143Z"
    }
   },
   "id": "864e5a9c0ee98257",
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### From/To HGVS\n",
    "This example will translate an HGVS variant to VRS using the *AlleleTranslator* *translate_from* method. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/989417/?oq=989417)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "87ddbc19e906119f"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA._Dy-JtNisSoxwmkOonFAjywkrVrtBz1d',\n 'type': 'Allele',\n 'digest': '_Dy-JtNisSoxwmkOonFAjywkrVrtBz1d',\n 'location': {'id': 'ga4gh:SL.ggzAgF2Nwx-oLD2NFXtSX4lEjtVN-YPK',\n  'type': 'SequenceLocation',\n  'digest': 'ggzAgF2Nwx-oLD2NFXtSX4lEjtVN-YPK',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.dLZ15tNO1Ur0IcGjwc3Sdi_0A6Yf4zm7'},\n  'start': 61860482,\n  'end': 61863459},\n 'state': {'type': 'LiteralSequenceExpression',\n  'sequence': 'TGCGCCACCACGCCCAGCTAATTTTGGATTTTTAG'}}"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_translator = AlleleTranslator(data_proxy=seqrepo_dataproxy)\n",
    "allele = allele_translator.translate_from(\"NC_000017.11:g.61860483_61863459delinsTGCGCCACCACGCCCAGCTAATTTTGGATTTTTAG\", \"hgvs\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:41.469507Z",
     "start_time": "2024-04-19T15:29:39.132283Z"
    }
   },
   "id": "925b01dd4764ed33",
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "source": [
    "The output from above is the VRS representation of the *Allele*. Using the *AlleleTranslator* *translate_to* method we can get back to the HGVS representation."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "f3951db9d1a1b833"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "['NC_000017.11:g.61860483_61863459delinsTGCGCCACCACGCCCAGCTAATTTTGGATTTTTAG']"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_translator.translate_to(allele, \"hgvs\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:43.260877Z",
     "start_time": "2024-04-19T15:29:41.471715Z"
    }
   },
   "id": "722919c1d8cdd83b",
   "execution_count": 5
  },
  {
   "cell_type": "markdown",
   "source": [
    "The AlleleTranslator class by default will use \"GRCh38\" as the default assembly when performing translation. But the actual assembly used for translation will be inferred from the reference sequence passed as part of the HGVS variant. A specific default assembly may be specified when creating an AlleleTranslator by passing in the keyword argument \"default_assembly_name\" with the assembly:\n",
    "> AlleleTranslator(data_proxy=seqrepo_dataproxy, default_assembly_name=\"GRCh37\")\n",
    "\n",
    "This example is using the GRCh37 representation of the variant. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/2461826) "
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "508d821d68360f36"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.4XLIRag1TPKSGV8YkIBtOJTOH5ETOg0p',\n 'type': 'Allele',\n 'digest': '4XLIRag1TPKSGV8YkIBtOJTOH5ETOg0p',\n 'location': {'id': 'ga4gh:SL.GcUa9awkL-Tw3vwvnKO19bO_DpatpIbr',\n  'type': 'SequenceLocation',\n  'digest': 'GcUa9awkL-Tw3vwvnKO19bO_DpatpIbr',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.XOgHwwR3Upfp5sZYk6ZKzvV25a4RBVu8'},\n  'start': 24379391,\n  'end': 24379392},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'T'}}"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = allele_translator.translate_from(\"NC_000022.10:g.24379392C>T\", \"hgvs\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:43.776563Z",
     "start_time": "2024-04-19T15:29:43.263138Z"
    }
   },
   "id": "ef3ba37deafba7ac",
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### From/To SPDI\n",
    "Example of translation a SPDI representation of a variant to and from VRS. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/2161661)."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "728fbf740565a801"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.VC-fz_GLT49WeUj2umwzGPxgldJHsq85',\n 'type': 'Allele',\n 'digest': 'VC-fz_GLT49WeUj2umwzGPxgldJHsq85',\n 'location': {'id': 'ga4gh:SL.mjTPtghQ7gXU8J-S7q1c8ksTaVRTADXn',\n  'type': 'SequenceLocation',\n  'digest': 'mjTPtghQ7gXU8J-S7q1c8ksTaVRTADXn',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.eK4D2MosgK_ivBkgi6FVPg5UXs1bYESm'},\n  'start': 95116662,\n  'end': 95116663},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'A'}}"
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = allele_translator.translate_from(\"NC_000014.9:95116662:G:A\",\"spdi\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:44.243901Z",
     "start_time": "2024-04-19T15:29:43.778734Z"
    }
   },
   "id": "43873e55f82d10a0",
   "execution_count": 7
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "['NC_000014.9:95116662:1:A']"
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele_translator.translate_to(allele, \"spdi\")"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:44.248908Z",
     "start_time": "2024-04-19T15:29:44.245300Z"
    }
   },
   "id": "dd695b3dfc14a83e",
   "execution_count": 8
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### From Beacon (VCF-like)\n",
    "For variants represented in the Beacon nomenclature, the *AlleleTranslator* currently only supports *translate_from* to convert to VRS. *translate_to* is not yet supported."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "4e0911a7694a060"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.GJ2JySBMXePcV2yItyvCfbGBUoawOBON',\n 'type': 'Allele',\n 'digest': 'GJ2JySBMXePcV2yItyvCfbGBUoawOBON',\n 'location': {'id': 'ga4gh:SL.28YsnRvD40gKu1x3nev0gRzRz-5OTlpS',\n  'type': 'SequenceLocation',\n  'digest': '28YsnRvD40gKu1x3nev0gRzRz-5OTlpS',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ._0wi-qoDrvram155UmcSC-zA5ZK4fpLT'},\n  'start': 32936731,\n  'end': 32936732},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'C'}}"
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = allele_translator.translate_from(\"13 : 32936732 G > C\", \"beacon\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:44.780851Z",
     "start_time": "2024-04-19T15:29:44.250188Z"
    }
   },
   "id": "57f54e6c3854a48f",
   "execution_count": 9
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### From gnomAD style VCF\n",
    "For variants represented in the gnomad nomenclature, the *AlleleTranslator* currently only supports *translate_from* to convert to VRS. *translate_to* is not yet supported.\n",
    "This variant can be viewed in [gnomAD](https://gnomad.broadinstitute.org/variant/1-55051215-G-GA)"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c4a6abfd56b8fa1e"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:VA.FgKdbB-uC6xIU2j8pVlCw04f8KsJUBcg',\n 'type': 'Allele',\n 'digest': 'FgKdbB-uC6xIU2j8pVlCw04f8KsJUBcg',\n 'location': {'id': 'ga4gh:SL.N8koWyjagSVChg4LKVmq9XepNlOrIPt6',\n  'type': 'SequenceLocation',\n  'digest': 'N8koWyjagSVChg4LKVmq9XepNlOrIPt6',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.Ya6Rs7DHhDeg7YaOSg1EoNi3U_nQ9SvO'},\n  'start': 55051215,\n  'end': 55051215},\n 'state': {'type': 'LiteralSequenceExpression', 'sequence': 'A'}}"
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# allele_translator = AlleleTranslator(data_proxy=seqrepo_dataproxy, default_assembly_name=\"GRCh37\")\n",
    "allele = allele_translator.translate_from(\"1-55051215-G-GA\", \"gnomad\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-19T15:29:45.850106Z",
     "start_time": "2024-04-19T15:29:44.782546Z"
    }
   },
   "id": "7868c365e327d995",
   "execution_count": 10
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
