{
 "cells": [
  {
   "cell_type": "markdown",
   "source": [
    "# 5 Exploring the CnvTranslator\n",
    "The vrs-python model supports two classes of copy number variation: \n",
    "* CopyNumberChange -  an assessment of loss or gain relative to a location within a gene or system, where loss or gain is represented by the following \"efo\" ontology codes:\n",
    "    * efo:0030064 - regional base ploidy\n",
    "    * efo:0030067 - loss\n",
    "    * efo:0030068 - low-level loss\n",
    "    * efo:0030069 - complete genomic loss\n",
    "    * efo:0030070 - gain\n",
    "    * efo:0030071 - low level gain\n",
    "    * efo:0030072 - high-level gain\n",
    "    * efo:0020073 - high-level loss \n",
    "* CopyNumberCount - an absolute count of discrete copies of a location within a gene or system\n",
    "For the CnvTranslator, only HGVS nomenclature is used to describe the variation."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "c940e9d78bc6e98a"
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Step 1 - Setup Data Proxy Access\n",
    "The *DataProxy* provides access to sequence references."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ac31eec4a405b218"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "from ga4gh.vrs.dataproxy import create_dataproxy\n",
    "seqrepo_rest_service_url = \"seqrepo+https://services.genomicmedlab.org/seqrepo\"\n",
    "seqrepo_dataproxy = create_dataproxy(uri=seqrepo_rest_service_url)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:27.328387Z",
     "start_time": "2024-04-18T20:25:27.194307Z"
    }
   },
   "id": "b7b0c4864ad5f9dd",
   "execution_count": 1
  },
  {
   "cell_type": "markdown",
   "source": [
    "Import the *CnvTranslator* class."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "27a861e38d55ea44"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "from ga4gh.vrs.extras.translator import CnvTranslator"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:27.379097Z",
     "start_time": "2024-04-18T20:25:27.329523Z"
    }
   },
   "id": "9fe4f25508590533",
   "execution_count": 2
  },
  {
   "cell_type": "markdown",
   "source": [
    "The UTA server is required in the environment since we are translating from/to HGVS."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "e9312b31a06c98e0"
  },
  {
   "cell_type": "code",
   "outputs": [],
   "source": [
    "import os\n",
    "os.environ[\"UTA_DB_URL\"] = \"postgresql://anonymous:anonymous@uta.biocommons.org:5432/uta/uta_20210129b\""
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:27.381511Z",
     "start_time": "2024-04-18T20:25:27.379793Z"
    }
   },
   "id": "71a288dff7b87f1",
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Step 2 - CopyNumberChange examples"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6c8160cff22db940"
  },
  {
   "cell_type": "markdown",
   "source": [
    "This example depicts a *CopyNumberChange* representing a deletion, or copy number loss. The \"efo\" ontology code specifying the type of copy number change is passed as a keyword argument \"copy_change\" to *translate_from*. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/984438)."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6a2c1c355bbcc494"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:CX.0M5VkV5v504_laQURFMEsqzZGcOF9YEw',\n 'type': 'CopyNumberChange',\n 'digest': '0M5VkV5v504_laQURFMEsqzZGcOF9YEw',\n 'location': {'id': 'ga4gh:SL.GSJAEJXFDz7Nq6VlJj5NTEku48MmteUU',\n  'type': 'SequenceLocation',\n  'digest': 'GSJAEJXFDz7Nq6VlJj5NTEku48MmteUU',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.eK4D2MosgK_ivBkgi6FVPg5UXs1bYESm'},\n  'start': 45002866,\n  'end': 45015056},\n 'copyChange': 'efo:0030067'}"
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cnv_translator = CnvTranslator(data_proxy=seqrepo_dataproxy)\n",
    "allele = cnv_translator.translate_from(\"NC_000014.9:g.45002867_45015056del\", \"hgvs\", copy_change=\"efo:0030067\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:28.974970Z",
     "start_time": "2024-04-18T20:25:27.382082Z"
    }
   },
   "id": "c8a6400cb8605b0d",
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "source": [
    "This example depicts a CopyNumberChange* representing a duplication, or copy number gain. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/549625)."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "6fc8dfa1340e010d"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:CX.0BN4vrqPrLPAZYsQEAPnG4IS8AYeBGe1',\n 'type': 'CopyNumberChange',\n 'digest': '0BN4vrqPrLPAZYsQEAPnG4IS8AYeBGe1',\n 'location': {'id': 'ga4gh:SL.tydo6UFL8Y60L5Me3k8AJfljURO9vYn9',\n  'type': 'SequenceLocation',\n  'digest': 'tydo6UFL8Y60L5Me3k8AJfljURO9vYn9',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI'},\n  'start': 75502957,\n  'end': 76045032},\n 'copyChange': 'efo:0030070'}"
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = cnv_translator.translate_from(\"NC_000009.12:g.75502958_76045032dup\", \"hgvs\", copy_change=\"efo:0030070\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:29.076267Z",
     "start_time": "2024-04-18T20:25:28.976791Z"
    }
   },
   "id": "f4efc189d53d7000",
   "execution_count": 5
  },
  {
   "cell_type": "markdown",
   "source": [
    "#### Step 3 - CopyNumberCount examples"
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "aaf041e4b83301cd"
  },
  {
   "cell_type": "markdown",
   "source": [
    "This example depicts a *CopyNumberCount* with a copy number gain. With copy number count variation, the \"copies\" keyword argument is passed to *translate_from* with the appropriate \"efo\" ontology code. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/2579174/)."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "ff76cb8a2f1387a5"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:CN.O_QHImmfErh9jDFkJaypPPvUmnj7EM70',\n 'type': 'CopyNumberCount',\n 'digest': 'O_QHImmfErh9jDFkJaypPPvUmnj7EM70',\n 'location': {'id': 'ga4gh:SL.hBVWalem_rNclxjmUuT9CHbEGCdlqW9L',\n  'type': 'SequenceLocation',\n  'digest': 'hBVWalem_rNclxjmUuT9CHbEGCdlqW9L',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.HxuclGHh0XCDuF8x6yQrpHUBL7ZntAHc'},\n  'start': 85623,\n  'end': 57073230},\n 'copies': 3}"
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = cnv_translator.translate_from(\"NC_000004.12:g.85624_57073230dup\", \"hgvs\", copies=\"3\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:29.187023Z",
     "start_time": "2024-04-18T20:25:29.077128Z"
    }
   },
   "id": "f057e93172e97a88",
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "source": [
    "This example depicts a *CopyNumberCount* with a copy number loss. This variant can be viewed in [ClinVar](https://www.ncbi.nlm.nih.gov/clinvar/variation/2579226/)."
   ],
   "metadata": {
    "collapsed": false
   },
   "id": "41df40dd67cb1009"
  },
  {
   "cell_type": "code",
   "outputs": [
    {
     "data": {
      "text/plain": "{'id': 'ga4gh:CN.WDzlT9oUq4IcQrVRWGH0dZnARnFBotCS',\n 'type': 'CopyNumberCount',\n 'digest': 'WDzlT9oUq4IcQrVRWGH0dZnARnFBotCS',\n 'location': {'id': 'ga4gh:SL.H1Zh5xdBqamBjwVE9orWdY_uBkpEMH1V',\n  'type': 'SequenceLocation',\n  'digest': 'H1Zh5xdBqamBjwVE9orWdY_uBkpEMH1V',\n  'sequenceReference': {'type': 'SequenceReference',\n   'refgetAccession': 'SQ.5ZUqxCmDDgN4xTRbaSjN8LwgZironmB8'},\n  'start': 46111352,\n  'end': 46119948},\n 'copies': 1}"
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "allele = cnv_translator.translate_from(\"NC_000021.9:g.46111353_46119948del\", \"hgvs\", copies=\"1\")\n",
    "allele.model_dump(exclude_none=True)"
   ],
   "metadata": {
    "collapsed": false,
    "ExecuteTime": {
     "end_time": "2024-04-18T20:25:29.276073Z",
     "start_time": "2024-04-18T20:25:29.187923Z"
    }
   },
   "id": "412feaeba95751e7",
   "execution_count": 7
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
