{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 209,
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "from timeit import default_timer as timer\n",
    "\n",
    "import ga4gh.vrs._internal.models\n",
    "import ga4gh.core._internal.identifiers\n",
    "import ga4gh.core._internal.pydantic \n",
    "import importlib\n",
    "importlib.reload(ga4gh.vrs._internal.models)\n",
    "importlib.reload(ga4gh.core._internal.identifiers)\n",
    "importlib.reload(ga4gh.core._internal.pydantic)\n",
    "\n",
    "from ga4gh.core._internal.pydantic import (\n",
    "    is_curie_type,\n",
    "    pydantic_copy,\n",
    "    is_pydantic_custom_str_type\n",
    ")\n",
    "from ga4gh.vrs._internal.models import (\n",
    "    Allele,\n",
    "    Haplotype,\n",
    "    Genotype,\n",
    "    GenotypeMember\n",
    ")\n",
    "from ga4gh.core._internal.identifiers import (\n",
    "    ga4gh_serialize,\n",
    "    ga4gh_identify,\n",
    "    ga4gh_digest,\n",
    "    sha512t24u,\n",
    "    identify_all,\n",
    "    collapse_identifiable_values,\n",
    "    replace_with_digest\n",
    ")\n",
    "def pretty_print(d: dict):\n",
    "    print(json.dumps(d, indent=2))"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "TODO\n",
    "\n",
    "- Update pydantic.py helper functions to work for classes in models.py (if those functions are needed)\n",
    "- Update Translator _from* functions to work for the simple Allele in test_vrs.py (also below in allele_dict)\n",
    "- Update the ga4gh_serialize and ga4gh_digest functions to work with Pydantic classes and the ga4ghDigest field in VRS 2.0\n",
    "- Update test in test_vrs.py to work for new serialize and digest functions, for update expected values for VRS 2.model\n",
    "- location start/end validation in Pydantic class\n",
    "- IRI validation in Pydantic class (curie, compacted identifier, or fully qualified IRI according to RFCs)\n",
    "\n",
    "\n",
    "VRS 2.0 merge issues with pydantic classes\n",
    "ga4gh.vrs.extras.localizer. Line 56, refers to _value. Line 49 refers to _id\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 210,
   "metadata": {},
   "outputs": [],
   "source": [
    "allele_dict = {\n",
    "    'location': {\n",
    "        'end': 55181320,\n",
    "        'start': 55181319,\n",
    "        'sequence': {\n",
    "            'type': 'SequenceReference',\n",
    "            'refgetAccession': 'SQ.F-LrLMe1SRpfUZHkQmvkVKFEGaoDeHul'\n",
    "        },\n",
    "        # 'sequence_id': 'ga4gh:SQ.F-LrLMe1SRpfUZHkQmvkVKFEGaoDeHul',\n",
    "        'type': 'SequenceLocation'\n",
    "    },\n",
    "    'state': {\n",
    "        'sequence': 'T',\n",
    "        'type': 'LiteralSequenceExpression'\n",
    "    },\n",
    "    'type': 'Allele'\n",
    "}\n",
    "\n",
    "allele = Allele(**allele_dict)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 211,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"location\": {\n",
      "    \"type\": \"SequenceLocation\",\n",
      "    \"start\": 55181319,\n",
      "    \"end\": 55181320,\n",
      "    \"sequence\": {\n",
      "      \"refgetAccession\": \"SQ.F-LrLMe1SRpfUZHkQmvkVKFEGaoDeHul\",\n",
      "      \"type\": \"SequenceReference\",\n",
      "      \"digest\": \"OFEyBMeo55q3QRrxAY5FiDqnkdyf0GTV\"\n",
      "    },\n",
      "    \"digest\": \"X0qrF7RfZxGIVIOddTYooZ_23D9mw6p6\"\n",
      "  },\n",
      "  \"state\": {\n",
      "    \"type\": \"LiteralSequenceExpression\",\n",
      "    \"sequence\": \"T\"\n",
      "  },\n",
      "  \"type\": \"Allele\",\n",
      "  \"digest\": \"oFiLzDh37SoecjP7dceRaUfVlh32NnCg\"\n",
      "}\n"
     ]
    }
   ],
   "source": [
    "allele_identified = identify_all(allele)\n",
    "pretty_print(allele_identified)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 212,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Location serialized: b'{\"end\":55181320,\"sequence\":\"OFEyBMeo55q3QRrxAY5FiDqnkdyf0GTV\",\"start\":55181319,\"type\":\"SequenceLocation\"}'\n",
      "Location digest: X0qrF7RfZxGIVIOddTYooZ_23D9mw6p6\n",
      "Location digest: X0qrF7RfZxGIVIOddTYooZ_23D9mw6p6\n",
      "Allele serialized: b'{\"location\":\"X0qrF7RfZxGIVIOddTYooZ_23D9mw6p6\",\"state\":{\"sequence\":\"T\",\"type\":\"LiteralSequenceExpression\"},\"type\":\"Allele\"}'\n",
      "Allele digest: oFiLzDh37SoecjP7dceRaUfVlh32NnCg\n",
      "Allele digest: oFiLzDh37SoecjP7dceRaUfVlh32NnCg\n"
     ]
    }
   ],
   "source": [
    "location_serialized = ga4gh_serialize(allele.location)\n",
    "print(\"Location serialized: \" + str(location_serialized))\n",
    "print(\"Location digest: \" + str(sha512t24u(location_serialized)))\n",
    "print(\"Location digest: \" + str(ga4gh_digest(allele.location)))\n",
    "\n",
    "allele_serialized = ga4gh_serialize(allele)\n",
    "print(\"Allele serialized: \" + str(allele_serialized))\n",
    "print(\"Allele digest: \" + str(sha512t24u(allele_serialized)))\n",
    "print(\"Allele digest: \" + str(ga4gh_digest(allele)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 213,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Average duration: 0.0001394366589956917\n"
     ]
    }
   ],
   "source": [
    "ct = 1000\n",
    "start = timer()\n",
    "for i in range(ct):\n",
    "    digest = ga4gh_digest(allele)\n",
    "end = timer()\n",
    "print(\"Average duration: \" + str((end - start) / ct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 214,
   "metadata": {},
   "outputs": [],
   "source": [
    "from ga4gh.vrs.extras.translator import Translator\n",
    "from ga4gh.vrs.dataproxy import SeqRepoDataProxy\n",
    "from biocommons.seqrepo import SeqRepo\n",
    "data_proxy = SeqRepoDataProxy(SeqRepo(\"/Users/kferrite/dev/biocommons.seqrepo/seqrepo/2021-01-29\"))\n",
    "translator = Translator(data_proxy=data_proxy)\n",
    "# translator._from_beacon(\"13 : 32936732 G > C\")\n",
    "# data_proxy.get_metadata('GRCh38:13')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 215,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"type\": \"Haplotype\",\n",
      "  \"members\": [\n",
      "    {\n",
      "      \"type\": \"Allele\",\n",
      "      \"location\": {\n",
      "        \"type\": \"SequenceLocation\",\n",
      "        \"sequence\": {\n",
      "          \"type\": \"SequenceReference\",\n",
      "          \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
      "        },\n",
      "        \"start\": 128325834,\n",
      "        \"end\": 128325835\n",
      "      },\n",
      "      \"state\": {\n",
      "        \"type\": \"LiteralSequenceExpression\",\n",
      "        \"sequence\": \"T\"\n",
      "      }\n",
      "    },\n",
      "    {\n",
      "      \"type\": \"Allele\",\n",
      "      \"location\": {\n",
      "        \"type\": \"SequenceLocation\",\n",
      "        \"sequence\": {\n",
      "          \"type\": \"SequenceReference\",\n",
      "          \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
      "        },\n",
      "        \"start\": 128325809,\n",
      "        \"end\": 128325810\n",
      "      },\n",
      "      \"state\": {\n",
      "        \"type\": \"LiteralSequenceExpression\",\n",
      "        \"sequence\": \"T\"\n",
      "      }\n",
      "    }\n",
      "  ]\n",
      "}\n",
      "Haplotype serialized: b'{\"members\":[\"WzQqNgeHveA5UCbiTWCivrVQEaSTKT5e\",\"iHLnKCHfdlR6KB_OXLkzqCf86WGH91dq\"],\"type\":\"Haplotype\"}'\n",
      "Haplotype digest: i7MKrP2OU1nH0WQfwc4W_8DWVU0bac8X\n",
      "Average duration: 0.0003378258219745476\n"
     ]
    }
   ],
   "source": [
    "# genotype \n",
    "# https://www.ncbi.nlm.nih.gov/clinvar/variation/431013/\n",
    "#   haplotype\n",
    "#   https://www.ncbi.nlm.nih.gov/clinvar/variation/431012/\n",
    "#     allele\n",
    "#     https://www.ncbi.nlm.nih.gov/clinvar/variation/383650/\n",
    "#     https://www.ncbi.nlm.nih.gov/clinvar/variation/417816/\n",
    "# TODO other simple allele in genotype\n",
    "\n",
    "# NC_000009.12:g.128325835C>T\n",
    "allele_383650_dict = {\n",
    "    \"type\": \"Allele\",\n",
    "    \"location\": {\n",
    "        \"type\": \"SequenceLocation\",\n",
    "        \"sequence\": {\n",
    "            \"type\": \"SequenceReference\",\n",
    "            \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
    "        },\n",
    "        \"start\": 128325834,\n",
    "        \"end\": 128325835\n",
    "    },\n",
    "    \"state\": {\n",
    "        \"type\": \"LiteralSequenceExpression\",\n",
    "        \"sequence\": \"T\"\n",
    "    }\n",
    "}\n",
    "allele_417816_dict = {\n",
    "    \"type\": \"Allele\",\n",
    "    \"location\": {\n",
    "        \"type\": \"SequenceLocation\",\n",
    "        \"sequence\": {\n",
    "            \"type\": \"SequenceReference\",\n",
    "            \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
    "        },\n",
    "        \"start\": 128325809,\n",
    "        \"end\": 128325810\n",
    "    },\n",
    "    \"state\": {\n",
    "    \"type\": \"LiteralSequenceExpression\",\n",
    "    \"sequence\": \"T\"\n",
    "    }\n",
    "}\n",
    "allele_383650 = Allele(**allele_383650_dict)\n",
    "allele_417816 = Allele(**allele_417816_dict)\n",
    "\n",
    "haplotype_431012 = Haplotype(**{\n",
    "    \"members\": [allele_383650, allele_417816]\n",
    "})\n",
    "pretty_print(haplotype_431012.model_dump(exclude_none=True))\n",
    "haplotype_431012_serialized = ga4gh_serialize(haplotype_431012)\n",
    "print(f\"Haplotype serialized: {haplotype_431012_serialized}\")\n",
    "print(f\"Haplotype digest: {ga4gh_digest(haplotype_431012)}\")\n",
    "\n",
    "\n",
    "ct = 1000\n",
    "start = timer()\n",
    "for i in range(ct):\n",
    "    digest = ga4gh_digest(haplotype_431012)\n",
    "end = timer()\n",
    "print(\"Average duration: \" + str((end - start) / ct))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 216,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{\n",
      "  \"type\": \"Genotype\",\n",
      "  \"members\": [\n",
      "    {\n",
      "      \"type\": \"GenotypeMember\",\n",
      "      \"count\": 1,\n",
      "      \"variation\": {\n",
      "        \"type\": \"Haplotype\",\n",
      "        \"members\": [\n",
      "          {\n",
      "            \"type\": \"Allele\",\n",
      "            \"location\": {\n",
      "              \"type\": \"SequenceLocation\",\n",
      "              \"sequence\": {\n",
      "                \"type\": \"SequenceReference\",\n",
      "                \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
      "              },\n",
      "              \"start\": 128325834,\n",
      "              \"end\": 128325835\n",
      "            },\n",
      "            \"state\": {\n",
      "              \"type\": \"LiteralSequenceExpression\",\n",
      "              \"sequence\": \"T\"\n",
      "            }\n",
      "          },\n",
      "          {\n",
      "            \"type\": \"Allele\",\n",
      "            \"location\": {\n",
      "              \"type\": \"SequenceLocation\",\n",
      "              \"sequence\": {\n",
      "                \"type\": \"SequenceReference\",\n",
      "                \"refgetAccession\": \"ga4gh:SQ.KEO-4XBcm1cxeo_DIQ8_ofqGUkp4iZhI\"\n",
      "              },\n",
      "              \"start\": 128325809,\n",
      "              \"end\": 128325810\n",
      "            },\n",
      "            \"state\": {\n",
      "              \"type\": \"LiteralSequenceExpression\",\n",
      "              \"sequence\": \"T\"\n",
      "            }\n",
      "          }\n",
      "        ]\n",
      "      }\n",
      "    }\n",
      "  ],\n",
      "  \"count\": 1\n",
      "}\n",
      "Genotype serialized: b'{\"count\":1,\"members\":[{\"count\":1,\"type\":\"GenotypeMember\",\"variation\":\"i7MKrP2OU1nH0WQfwc4W_8DWVU0bac8X\"}],\"type\":\"Genotype\"}'\n",
      "Genotype digest: QwK1T7G4p-y2dGsj4QqmvqLMf1dYRmnZ\n",
      "Average duration: 0.00044584115399629807\n"
     ]
    }
   ],
   "source": [
    "genotype_431013_dict = {\n",
    "    \"type\": \"Genotype\",\n",
    "    \"count\": 1,\n",
    "    \"members\": [\n",
    "        {\n",
    "            \"type\": \"GenotypeMember\",\n",
    "            \"variation\": haplotype_431012,\n",
    "            \"count\": 1\n",
    "        }\n",
    "    ]\n",
    "}\n",
    "genotype_431013 = Genotype(**genotype_431013_dict)\n",
    "\n",
    "pretty_print(genotype_431013.model_dump(exclude_none=True))\n",
    "genotype_431013_serialized = ga4gh_serialize(genotype_431013)\n",
    "print(f\"Genotype serialized: {genotype_431013_serialized}\")\n",
    "print(f\"Genotype digest: {ga4gh_digest(genotype_431013)}\")\n",
    "\n",
    "ct = 1000\n",
    "start = timer()\n",
    "for i in range(ct):\n",
    "    digest = ga4gh_digest(genotype_431013)\n",
    "end = timer()\n",
    "print(\"Average duration: \" + str((end - start) / ct))\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.3"
  },
  "orig_nbformat": 4
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
